timers/timers_test.go

```diff
@@ -0,0 +1,419 @@
+package timers
+
+import (
+	"math"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+	"unsafe"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/stretchr/testify/assert"
+)
+
+// O(n), only for use in tests.
+func (t *Timer) len() int {
+	n := 0
+	for el := t; el != nil; el = el.next {
+		n++
+	}
+	return n
+}
+
+func fakeWheel(tick time.Duration) (*Wheel, *clock.Mock) {
+	clock := clock.NewMock()
+	w := newWheel(tick, 2*time.Minute, clock)
+	w.start()
+	return w, clock
+}
+
+func newBucketList() *bucketList {
+	return &bucketList{
+		mask:    uint64(1<<3 - 1),
+		buckets: make([]bucket, 1<<3),
+	}
+}
+
+func assertQueuedTimers(t testing.TB, bs *bucketList, bucketIdx, expected int) {
+	b := &bs.buckets[bucketIdx]
+	b.Lock()
+	defer b.Unlock()
+	assert.Equal(
+		t,
+		expected,
+		b.timers.len(),
+		"Unexpected number of counters in bucket %v.", bucketIdx,
+	)
+}
+
+func fakeWork() {} // non-nil func to schedule
+
+func randomTimeouts(n int, max time.Duration) []time.Duration {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	ds := make([]time.Duration, n)
+	for i := 0; i < n; i++ {
+		ds[i] = time.Duration(r.Int63n(int64(max)))
+	}
+	return ds
+}
+
+func TestTimersLinkedListPushOneNils(t *testing.T) {
+	var root *Timer
+	head := newTimer(0, nil)
+	root = root.pushOne(head)
+	assert.Equal(t, 1, root.len(), "Unexpected length after pushing one timer to nil root.")
+	assert.Equal(t, head, root, "Expected pushOne with nil receiver to return head.")
+	assert.Nil(t, root.next, "Expected one-node list to have a nil next link.")
+	assert.Equal(t, root, root.tail, "Expected head of single-node list to be its own tail.")
```

Comment:
Instead of just a `len` method, should have a method which does more validation that is useful for more tests,
```go
func assertLinkedList(t *testing.T, timer *Timer, expectedLen int) {
  if expectedLen == 0 {
    assert.Nil(t, timer, "Empty linked list should be nil")
    return
  }

  nodes := 0
  last := head
  for cur := head; cur != nil; cur = last.next {
    if cur != head {
      assert.Nil(t, cur.tail, "Only the head should have a tail set")
    }
    nodes++
    last = cur
  }

  // Verify tail is actually the last node.
  assert.Equal(t, head.tail, last, "Tail pointer is incorrect")
  assert.Equal(t, expectedLen, nodes, "Number of nodes is incorrect")
}
```

Then, the validation here would really only be:
```
var root *Timer
assertLinkedList(t, root, 0)

head := newTimer(0, nil)
root := root.pushOne(head)
assertLinkedList(t, root, 1)
assert.Equal(t, head, root, [...])

assert.Panics(...)
```

There's no need to verify the `next` and `tail` since the `assertLinkedList` will do that. Same is true for other tests.